extensions/vertx-keycloak/runtime/src/main/java/io/quarkus/vertx/keycloak/VertxOAuth2AuthenticationMechanism.java

```diff
@@ -0,0 +1,110 @@
+
+    private static final String BEARER = "Bearer";
+
+    private volatile String authServerURI;
+    private volatile OAuth2Auth auth;
+    private volatile KeycloakConfig config;
+
+    public KeycloakConfig getConfig() {
+        return config;
+    }
+
+    public VertxOAuth2AuthenticationMechanism setConfig(KeycloakConfig config) {
+        this.config = config;
+        return this;
+    }
+
+    public String getAuthServerURI() {
+        return authServerURI;
+    }
+
+    public VertxOAuth2AuthenticationMechanism setAuthServerURI(String authServerURI) {
+        this.authServerURI = authServerURI;
+        return this;
+    }
+
+    public OAuth2Auth getAuth() {
+        return auth;
+    }
+
+    public VertxOAuth2AuthenticationMechanism setAuth(OAuth2Auth auth) {
+        this.auth = auth;
+        return this;
+    }
+
+    @Override
+    public CompletionStage<SecurityIdentity> authenticate(RoutingContext context,
+            IdentityProviderManager identityProviderManager) {
+        // when the handler is working as bearer only, then the `Authorization` header is required
+
+        final HttpServerRequest request = context.request();
+        final String authorization = request.headers().get(HttpHeaders.AUTHORIZATION);
+
+        if (authorization == null) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        int idx = authorization.indexOf(' ');
+
+        if (idx <= 0) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        if (!BEARER.equalsIgnoreCase(authorization.substring(0, idx))) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        String token = authorization.substring(idx + 1);
+        return identityProviderManager.authenticate(new TokenAuthenticationRequest(new TokenCredential(token, "oauth2")));
+    }
+
+    @Override
+    public CompletionStage<Boolean> sendChallenge(RoutingContext context) {
+        context.response().setStatusCode(302);
+        context.response().headers().set(HttpHeaders.LOCATION, authURI(authServerURI));
+        return CompletableFuture.completedFuture(true);
+    }
+
+    private String authURI(String redirectURL) {
+        final JsonObject config = new JsonObject()
+                .put("state", redirectURL);
+
+        config.put("redirect_uri", authServerURI);
+
+        //        if (extraParams != null) {
+        //            config.mergeIn(extraParams);
+        //        }
+        //
+        //        if (scopes.size() > 0) {
+        //            JsonArray _scopes = new JsonArray();
+        //            // scopes are passed as an array because the auth provider has the knowledge on how to encode them
+        //            for (String authority : scopes) {
+        //                _scopes.add(authority);
+        //            }
+        //
+        //            config.put("scopes", _scopes);
+        //        }
```

Comment:
Do we need a TODO for that?

Reply:
I am not 100% sure, this is going to need some coordination with the Keycloak team to decide exactly what is needed for 1.0.

Reply:
@stuartwdouglas 
This code is required for the authorization code flow (and others) to work: here the adapter is about to redirect a user to the (KC) OIDC server to authenticate, once the user is authenticatad he/she will be redirected back with a token grant and the adapter code will exchange for the access token: the 'scopes'  is what the client/adapter wants to do with this token, example, read+write, or read only, etc, some custom values possible - it is like token roles